Episode 3 - connecting to MUSCLE (#2)
esciencecenter-digital-skills/model-coupling-workshop: episodes/03-muscle-connect.md

```python
    U = msg.data.array.copy()
    t_cur = msg.timestamp
```

::: callout
The msg.data attribute holds an object of type libmuscle.Grid, which holds a read-only NumPy array and optionally a list of index names. Here, we take the array and make a copy of it for U, so that we can modify U in our upcoming state update. Without calling `.copy()`, U would end up pointing to the same read-only array, and we’d get an error message if we tried to modify it.

The `timestamp` attribute is a double precision float containing the number of simulated (not wall-clock) seconds since the whole simulation started. 
:::::::::::

## State update loop
Last thing is a small change in our state update loop. Instead of checking whether the simulation time does not exceed `t_max`, we have to now add up the time when we started the submodel instance, since we are part of a larger simulation.

```python
    # begin state_update_loop

>         while t_cur + dt < msg.timestamp + t_max:

Author: @LourensVeen Isn't it more efficient to leave the state update loop as it was: `while t_cur + dt < t_max`
and make sure that the received timestamp is added to t_cur when a global time stamp is sent out? Is there a good reason to do this addition every time in the while loop?

Reviewer: Yes, `msg.timestamp` and `t_max` don't change while we're iterating, so we could do that addition once, after receiving `msg` and before starting the loop.

We cannot add the received timestamp to `t_cur` though, as both are absolute times. Every time we enter the state update loop, we need to run from `msg.timestamp` to `msg.timestamp + t_max`. Does that make sense?

Author: I changed it in my latest commit to add the timestamp in the initialisation already, the reuse loop then stays as simple as possible and unchanged from the code we start with. See this commit: https://github.com/esciencecenter-digital-skills/model-coupling-workshop/pull/2/commits/eb0af7e9d0ec6b829b6f585203708951b37b8f5a

Reviewer: Okay, yes this looks good.

The pedantic hairsplitter in the back of my mind may want to change t_max to t_end at some point, but it would probably be good to keep this and the MUSCLE3 docs in sync, and it can wait. So let's keep it at this, nice job!

- Learn to connect your own model to the MUSCLE3 library

:::

## Introduction

MUSCLE3 is the third incarnation of the Multiscale Coupling Library and Environment. Its purpose is to make creating coupled multiscale simulations easy, and to then enable efficient Uncertainty Quantification of such models using advanced semi-intrusive algorithms.

MUSCLE3 uses the Multiscale Modelling and Simulation Language (MMSL) to describe the structure of a multiscale model and you will notice that the terminology in MUSCLE3 closely links to what you have learned in the previous episode about MMSF.

MUSCLE3 consist of three components:

**libmuscle** is a library that is used to connect component intances (e.g new or existing submodels) to a coupled simulation.

**muscle_manager** is the central manager of MUSCLE3 that starts up submodels and coordinates their interaction.

**ymmsl-python** is a Python library that contains class definitions to represent an MMSL model description. A yMMSL file serves as the interface between us (humans) and the muscle manager and is meant to describe the multiscale simulation and tell the muscle manager what to do.

> We continue with the reaction-diffusion model example that we used the previous episode and in this episode we will connect an implementation of the reaction model to the MUSCLE3 library step by step.

Reviewer 2: ```suggestion
We continue with the reaction-diffusion model example that we used in the previous episode and in this episode we will connect an implementation of the reaction model to the MUSCLE3 library step by step.
```

> ## Disect your model

Reviewer 2: ```suggestion
## Dissect your model
```

In the previous episode we have discussed the Submodel Execution Loop (SEL) and the various operators that are associated with it.

::: challenge

> ## Exercise 1: Can you recognize the Submodel Execution Loop?

Reviewer 2: It could first be helpful to the learners if we introduce the (admittedly simple) reaction model - what is it actually modelling? (Or will this be in the earlier theory episode already?)

Author: That depends a bit on Lourens' theory section. I suppose it would fit in there quite well since the model gets introduced there. But maybe a short description/walk through the code would fit here. I will add it.

> Open the file called reaction.py in a text editor. Can you recognize the beginning and end of the four operators ($F_{init}$, $O_I$, $S$ and $O_F$ ) plus the state update loop in this submodel? Mark them with comments in the code (e.g. `# begin F_init`, 10 comments in total).

Reviewer 2: ```suggestion
Open the file called reaction.py in a text editor (see below for code). Can you recognize the beginning and end of the four operators ($F_{init}$, $O_I$, $S$ and $O_F$ ) plus the state update loop in this submodel? Mark these by placing the following 10 comments in the code:
- `# begin F_INIT`
- `# end F_INIT`
- `# begin O_I`
- `# end O_I`
- `# begin S`
- `# end S`
- `# begin state_update_loop`
- `# end state_update_loop`
- `# begin O_F`
- `# end O_F`
```

Reviewer 2: Feel free to reject this suggestion, but I think it's clearer in the exercise  description to explicitly list the 10 comments we want them to place in the code.

```python
def reaction(initial_state: np.array) -> np.array:
    """
    ...
    """
    t_max = 2.469136e-6
    dt = 2.469136e-8
    k = -4.05e4

    U = initial_state

    t_cur = 0
    while t_cur + dt < t_max:
        U += k * U * dt
        t_cur += dt

    return U
```

:::::: solution

## Solution

```python
def reaction(initial_state: np.array) -> np.array:
    """
    ...
    """
    # begin F_INIT
    t_max = 2.469136e-6
    dt = 2.469136e-8
    k = -4.05e4

    U = initial_state

    t_cur = 0
    # end F_INIT

    # begin state_update_loop
    while t_cur + dt < t_max:
        # begin O_I
        # end O_I

        # begin S
        U += k * U * dt
        t_cur += dt
        # end S
    # end state_update_loop

    # begin O_F
    return U
    # end O_F
```

> The Submodel Execution Loop specifies two operators within the state update loop. $O_I$ and comes before $S$ and you can use it to send a message to the outside world with (part of) the current state. In this case, the $O_I$ operator is empty; we are not sending or storing any intermediate states. The original type annotations in the function definition serve the same purpose but we will be losing these once we are done connecting the model to MUSCLE.

Reviewer 2: `$O_I$ and comes before`
Remove the "and"? Or was there another operator here?

> The Submodel Execution Loop specifies two operators within the state update loop. $O_I$ and comes before $S$ and you can use it to send a message to the outside world with (part of) the current state. In this case, the $O_I$ operator is empty; we are not sending or storing any intermediate states. The original type annotations in the function definition serve the same purpose but we will be losing these once we are done connecting the model to MUSCLE.

Author: ```suggestion
The Submodel Execution Loop specifies two operators within the state update loop. $O_I$ comes before $S$ and you can use it to send a message to the outside world with (part of) the current state. In this case, the $O_I$ operator is empty; we are not sending or storing any intermediate states. The original type annotations in the function definition serve the same purpose but we will be losing these once we are done connecting the model to MUSCLE.
```

:::::

:::

## The constructor

> To let a model communicate with the muscle manager, other submodels and the outside world we need to create a `libmuscle.Instance` object:

Reviewer 2: ```suggestion
To let a model communicate with the muscle manager, other submodels and the outside world, we need to create a `libmuscle.Instance` object:
```

:::::: solution

In this model, we need:

- a single input port that will receive an initial state (e.g. named "initial_state") at the beginning of the model run and it should be mapped to the $F_{init}$ operator
- a single output port that sends the final state (e.g. named "final_State") at the end of the reaction simulation to the rest of the simulation, it should be mapped to the $O_F$ operator
::::::

:::

In multiscale simulations, submodels often have to run multiple times, for instance because they are used as a micro model or because they are part of an ensemble that cannot be completely parallelised. In order to accomplish this, the entire submodel is wrapped into a loop. Exactly when this loop needs to end often depends on the behaviour of the whole model, and is not easy to determine beforehand, but fortunately MUSCLE will do that for us if we call the `libmuscle.Instance.reuse_instance()` method.

```python
    while instance.reuse_instance():
        # F_INIT
        # State update loop
        # O_F
```

::: challenge

> ## Exercise 2: Constructing your submodel

Author: ```suggestion
## Exercise 2: The MUSCLE constructor
```

Add the following to our reaction model code:

- a `libmuscle.Instance` object with the appropriate operator-port dictionary
- the reuse loop

:::: solution

> ```python

Reviewer 2: "Exercise 2: Constructing your submodel" suggests (to me) that you should have a working/finished submodel at the end of this exercise. I would therefore probably attempt to replace the `F_INIT` section with code that gets the values from `initial_state` and then get stuck. Is there a way we can clarify here that the "solution" will only be a work in progress, with further steps needed to finish the submodel?

Author: I solved this by renaming the exercise, let me know if you think this is enough.

from libmuscle import Instance
from ymmsl import Operator

def reaction(initial_state: np.array) -> np.array:
    """
    ...
    """
    instance = Instance({
            Operator.F_INIT: ['initial_state'],       # np.array
            Operator.O_F: ['final_state']})           # np.array

    while instance.reuse_instance():
        # begin F_INIT
        t_max = 2.469136e-6
        dt = 2.469136e-8
        k = -4.05e4

        U = initial_state

        t_cur = 0
        # end F_INIT

        # begin state_update_loop
        while t_cur + dt < t_max:
            # begin O_I
            # end O_I

            # begin S
            U += k * U * dt
            t_cur += dt
            # end S
        # end state_update_loop

        # begin O_F
        return U
        # end O_F
```

Note that the type of data that is sent is documented in a comment. This is obviously not required, but it makes life a lot easier if someone else needs to use the code or you haven’t looked at it for a while, so it’s highly recommended.

::::

:::

## Settings

> Next is the first part of the model, in which the model is initialised. Nearly every model needs some settings that define how it behaves (e.g. the size of a timestep or model specific parameters). We would like to control these settings centrally suing the MUSCLE manager when starting up the coupled simulation so we don't have to change it in our model code every time if we want to try a range of values for instance to perform a sensitivity analysis. We can use the `libmuscle.Instance.get_setting` function instead to ask the MUSCLE manager for the values. The values, or ranges of values, should be recorded in the configuration file which we will cover in the next chapter.

Reviewer 2: ```suggestion
Next is the first part of the model, in which the model is initialised. Nearly every model needs some settings that define how it behaves (e.g. the size of a timestep or model specific parameters). We would like to control these settings centrally using the MUSCLE manager when starting up the coupled simulation so we don't have to change it in our model code every time if we want to try a range of values (for instance, to perform a sensitivity analysis). We can use the `libmuscle.Instance.get_setting` function instead to ask the MUSCLE manager for the values. The values, or ranges of values, should be recorded in the configuration file which we will cover in the next chapter.
```

```python
    some_variable = instance.get_setting('variable_name', 'variable_type')
```

::: callout
The second argument, which specifies the expected type, is optional. If it is given, MUSCLE will check that the user specified a value of the correct type, and if not raise an exception.
:::

::: challenge

## Exercise 3: 

> In our example several settings have been hard-coded into the model:

Reviewer 2: ```suggestion
In our example, several settings have been hard-coded into the model:
```

> - the total simulation time to run this sub-model `tmax`

Reviewer 2: ```suggestion
- the total simulation time to run this sub-model, `tmax`
```

> - the time step to use `dt`

Reviewer 2: ```suggestion
- the time step to use, `dt`
```

> - and the model parameter `k`

Reviewer 2: ```suggestion
- and the model parameter, `k`
```

:::::: solution

```python
from libmuscle import Instance
from ymmsl import Operator

def reaction(initial_state: np.array) -> np.array:
    """
    ...
    """
    instance = Instance({
            Operator.F_INIT: ['initial_state'],       # np.array
            Operator.O_F: ['final_state']})           # np.array

    while instance.reuse_instance():
        # begin F_INIT
        t_max = instance.get_setting('t_max', 'float')
        dt = instance.get_setting('dt', 'float')
        k = instance.get_setting('k', 'float')

        U = initial_state

        t_cur = 0
        # end F_INIT

        # begin state_update_loop
        ...
```

Note that getting settings needs to happen within the reuse loop; doing it before can lead to incorrect results.

::::::
:::

## Receiving messages
Apart from settings, we can use the `libmuscle.Instance.receive` function to receive an initial state for this submodel on the `initial_state` port. Note that we have declared that port above, and declared it to be associated with the `F_INIT` operator. During `F_INIT`, messages can only be received, not sent, so that declaration makes `initial_state` a receiving port.

> The message that we will receive can contain several pieces of information. For now, we are interested in the `data` and `timestamp` attributes. We assume the data to be a grid of floats containing our initial state and the time stamp tells us the simulated time at which this state is valid. We can store receive a message and store the `data` and `timestamp` attributes in the following way:

Reviewer 2: ```suggestion
The message that we will receive can contain several pieces of information. For now, we are interested in the `data` and `timestamp` attributes. We assume the data to be a grid of floats containing our initial state and the time stamp tells us the simulated time at which this state is valid. We can receive a message and store the `data` and `timestamp` attributes in the following way:
```

?

```python
    msg = instance.receive('initial_state')

    data = msg.data.array.copy()
    timestamp = msg.timestamp
```

::: callout
The `msg.data` attribute holds an object of type `libmuscle.Grid`, which holds a read-only NumPy array and optionally a list of index names. Here, we take the array and make a copy of it for `data`, so that we can modify `data` in our upcoming state update. Without calling `.copy()`, the variable `data` would end up pointing to the same read-only array, and we would get an error message if we tried to modify it.

The `timestamp` attribute is a double precision float containing the number of simulated (not wall-clock) seconds since the whole simulation started. Since `t_cur` is assigned the value of the timestamp, we don't need to make a copy.
:::

## State update loop

> The actual state update happens in the operator $S$ in the Submodel Execution Loop, and in our example model, it is determined entirely by the current state. Since no information from outside is needed, we do not receive any messages, and in our `libmuscle.Instance` declaration above, we did not declare any ports associated with `Operator.S`. 

Reviewer 2: ```suggestion
The actual state update happens in the operator $S$ in the Submodel Execution Loop and, in our example model, it is determined entirely by the current state. Since no information from outside is needed, we do not receive any messages, and in our `libmuscle.Instance` declaration above, we did not declare any ports associated with `Operator.S`. 
```

The operator `O_I` provides for observations of intermediate states. In other words, here is where you can send a message to the outside world with (part of) the current state. In this case, the `O_I` operator is empty; we’re not sending anything.

> One thing we need to be aware of is time. Most models depend on time in some way and it is good to be aware of the various time frames. MUSCLE3 enables passing on timestamps 

Reviewer 2: I'm not really sure what "MUSCLE3 enables passing on timestamps" means

Author: Ah this was a line of thought that was lost at some point... I completed the sentence.

> One thing we need to be aware of is time. Most models depend on time in some way and it is good to be aware of the various time frames. MUSCLE3 enables passing on timestamps 

Author: ```suggestion
One thing we need to be aware of is time. Most models depend on time in some way and it is good to be aware of the various time frames. Libmuscles messages support passing on timestamps to be able to keep track of the global simulation time of the coupled model. It is easier to keep track of only one time frame and it is therefore good practice to add the timestamp that accompanied the initial state to the sub-model simulation time steps instead of starting at zero for every sub-model instance. 
```

:::::::::
::::::
:::

## Sending messages

To send a message we first have to construct a `libmuscle.Message` object containing the current simulation time and the current state. We will convert the `numpy.array` to a `libmuscle.Grid` object first.

::: callout
We convert our `numpy.array` explicitly into a `libmuscle.Grid object`, so that we can add the name of the dimensions. In our example case there is only one, and "x" is not very descriptive, so we could have also passed the array directly, in which case MUSCLE would have done the conversion for us and sent a `libmuscle.Grid` without index names automatically.

> Note that grids (and NumPy arrays) are much more efficient than lists and dictionaries. If you have a lot of data to send, then you should use those as much as possible. For example, a set of agents is best sent as a dictionary of 1D NumPy arrays, with one array/grid for each attribute. A list of dictionaries will be quite a bit slower and use a lot more memory, but it should work up to a million or so objects.

Reviewer 2: Will all participants know what is meant here by a "set of agents"?

Author: @raar1 Maybe we should remove this piece altogether. It doesn't really fit in this introduction tutorial to mention such a optimization detail.  Just remove line 341 altogether. Do you agree @LourensVeen ?
```suggestion
```

:::

> The optional second parameter is a second timestamp, which is only used in [more advance applications](https://muscle3.readthedocs.io/en/latest/tutorial.html#message-timestamps) and can be set to `None` here.

Reviewer 2: ```suggestion
The optional second parameter is a second timestamp, which is only used in [more advanced applications](https://muscle3.readthedocs.io/en/latest/tutorial.html#message-timestamps) and can be set to `None` here.
```

```python
from libmuscle import Grid, Message
my_message = Message(my_timestamp, None, Grid(my_state, ['x']))
```

> To send the message we specify the port on which to send which again needs to match the declaration.

Reviewer 2: ```suggestion
To send the message we specify the port on which to send, which again needs to match the declaration.
```

            Operator.O_F: ['final_state']})           # np.array

    while instance.reuse_instance():
        # begin F_INIT
        t_max = instance.get_setting('t_max', 'float')
        dt = instance.get_setting('dt', 'float')
        k = instance.get_setting('k', 'float')

        msg = instance.receive('initial_state')
        U = msg.data.array.copy()

        t_cur = msg.timestamp
        t_max = msg.timestamp + t_max
        # end F_INIT

        # begin state_update_loop
        while t_cur + dt < t_max:
            # begin O_I
            # end O_I

            # begin S
            U += k * U * dt
            t_cur += dt
            # end S
        # end state_update_loop

        # begin O_F
        final_message = Message(t_cur, None, Grid(U, ['x']))
        instance.send('final_state', final_message)
        # end O_F
```

::::::::::::
:::::::::::::

Right now your sub-model is ready to be used by the muscle manager. We will couple it to another submodel and configure it to run the coupled simulation in the next chapter.

::: keypoints

- MUSCLE3 implements the MMSF theory, using the same terminology and MMSL for configuration of the coupled simulation

> - First analyze your code to    find the Submodel Execution Loop structure

Reviewer 2: ```suggestion
- First analyze your code to find the Submodel Execution Loop structure
```
